src/ethereum/trace.py

```diff
@@ -0,0 +1,86 @@
+"""
+.. _trace:
+
+EVM Trace
+^^^^^^^^^
+
+.. contents:: Table of Contents
+    :backlinks: none
+    :local:
+
+Introduction
+------------
+
+Defines the functions required for creating evm traces during execution.
+"""
+
+from typing import Any
+
+
+def capture_tx_start(env: Any) -> None:
+    """
+    Capture the state at the beginning of a transaction.
+    """
+    pass
+
+
+def capture_tx_end(
+    env: Any, gas_used: int, output: bytes, has_erred: bool
+) -> None:
+    """
+    Capture the state at the end of a transaction.
+    """
+    pass
+
+
+def capture_precompile_start(evm: Any, address: Any) -> None:
```

Comment:
My only architectural concern is that with multiple functions like you have here, a tracing implementation cannot listen for _all_ events, only specific ones. So lets say we create a logging tool that prints to stdout each trace event. The logging tool doesn't understand the events, just prints out the arguments.

If we have to add a new trace event much later on, now we have to go and update every tool built on top of the specs because those tools have to inject a new function to get notified about those events.

If we were in Rust, I'd suggest something like:

```rust
enum TraceEvent {
    PrecompileStart(address: Address),
    PrecompileEnd,
    // ...
}

fn evm_trace(evm: Evm, event: TraceEvent) {
    // ...
}
```

In Python... it doesn't feel as nice, but maybe:

```python
@dataclass
class PrecompileStart:
    address: object

@dataclass
class PrecompileEnd:
    pass

TraceEvent = Union[PrecompileStart, PrecompileEnd, ...]

def evm_trace(evm: object, event: TraceEvent) -> None:
    pass
```

Is this too much overhead for a theoretical concern? Too many new types cluttering up the spec?

Reply:
Updated